pytorch_lightning/strategies/bagua.py

```diff
@@ -0,0 +1,239 @@
+        distributed training algorithms and system optimizations.
+
+        Arguments:
+            algorithm:  Distributed algorithm used to do the actual communication and update. Built-in algorithms
+                include "gradient_allreduce", "bytegrad", "decentralized", "low_precision_decentralized", "qadam" and
+                "async".
+            do_flatten: Whether to flatten the Bagua communication buckets. The flatten operation will reset data
+                pointer of bucket tensors so that they can use faster code paths. Default: ``True``.
+            kwargs: Additional arguments that will be passed to initialize the Bagua algorithm.
+        """
+        if not _BAGUA_AVAILABLE:
+            raise MisconfigurationException(
+                "To use the `BaguaStrategy`, you must have `Bagua` installed. Use `pip install bagua` to install it."
+            )
+
+        super().__init__(
+            accelerator=accelerator,
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+            checkpoint_io=checkpoint_io,
+            precision_plugin=precision_plugin,
+        )
+
+        self._bagua_algorithm = algorithm
+        self._bagua_do_flatten = do_flatten
+        self._bagua_kwargs = kwargs
+
+    def setup_environment(self) -> None:
+        # start the other scripts
+        if not self.cluster_environment.creates_processes_externally:  # type: ignore[union-attr]
+            self._call_children_scripts()
+
+        self.setup_distributed()
+
+    def setup_distributed(self) -> None:
+        reset_seed()
+
+        # determine which process we are and world size
+        self.set_world_ranks()
+
+        self._init_bagua_distributed()
+
+    def _init_bagua_distributed(self) -> None:
+
+        self._set_node_environment_variables()
+        log.info(
+            "initializing Bagua distributed: "
+            f"GLOBAL_RANK: {self.global_rank}, "
+            f"MEMBER: {self.global_rank + 1}/{self.world_size}"
+        )
+
+        # need to set device first before initialize Bagua distributed environment
+        torch.cuda.set_device(self.local_rank)
+
+        if not is_initialized():
+            bagua.init_process_group()
+
+    def _set_node_environment_variables(self) -> None:
+        os.environ["MASTER_ADDR"] = self.cluster_environment.main_address  # type: ignore[union-attr]
+        os.environ["MASTER_PORT"] = str(self.cluster_environment.main_port)  # type: ignore[union-attr]
+        os.environ["RANK"] = str(self.global_rank)
+        os.environ["NODE_RANK"] = str(self.node_rank)
+        os.environ["WORLD_SIZE"] = str(self.world_size)
+        os.environ["LOCAL_RANK"] = str(self.local_rank)
+
+    def _check_qadam_optimizer(self) -> None:
+
+        trainer = self.lightning_module.trainer
+        has_qadam_optimizer = any([isinstance(opt, QAdamOptimizer) for opt in trainer.optimizers])
+
```

Comment:
we shouldn't need this indirection now that the Strategy class owns the optimizers

Reply:
done